component/prometheus/interceptor.go

```diff
@@ -102,6 +103,13 @@ func (a *interceptappender) Append(ref storage.SeriesRef, l labels.Labels, t int
 		ref = storage.SeriesRef(a.ls.GetOrAddGlobalRefID(l))
 	}
 
+	if value.IsStaleNaN(v) {
+		a.ls.AddStaleMarker(uint64(ref), l)
+	} else {
+		// This feels excessive to call this each time.
+		a.ls.RemoveStaleMarker(uint64(ref))
+	}
```

Comment:
Hmm, we did appear sometimes to run into bottleneck (WAL delay was growing) without hitting any obvious resource limitation. Could it be contention on locks like this? (of course not here, cause it's new, but similar calls to labelstore).

So I think there are two main questions I have here:
* Are there metrics that we could use to measure the overhead this could introduce?
* Could we test this branch out in dev first?

If we find there is a bottleneck here, we perhaps should consider having using a striped implementation of labelcache, where we have N locks and use ref hash to pick which lock is needed 🤔 increasing effectively number of possible concurrent operations on the labelstore. Would hopefully be able to find an existing library for it?

Reply:
The locks do show up on profiles, I though about using a bucketed approach but hesitant to add more complexity to the store. I would be more pro seeing how this behaves and then adjusting again in a more focused PR. This should be less overhead than the other calls at least.

Reply:
This ended up not having a significant impact on cpu after running it for 12 hours.

Reply:
Amazing, thanks for testing!

Reply:
Going to update my comment :) so its clear we tested it